sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -491,7 +492,10 @@ def get(self, instruction_id, bundle_descriptor_id):
     # Make sure we instantiate the processor while not holding the lock.
     processor = bundle_processor.BundleProcessor(
         self.runner_capabilities,
-        self.fns[bundle_descriptor_id],
+        # Reduce risks of concurrent modifications of the same protos
+        # captured in bundle descriptor when the same bundle desciptor is used
+        # in different instructions.
+        copy.deepcopy(self.fns[bundle_descriptor_id]),
```

Comment:
Not sure if it would be safer to use Message generated CopyFrom method instead. Unclear to me if deepcopy would do something odd with lower-level upb impl.
deepcopy isn't mentioned here https://protobuf.dev/reference/python/python-generated/

Reply:
deepcopy (impl: https://github.com/python/cpython/blob/9263173280d7ce949911965efa5b745287c581b2/Lib/copy.py#L118) should defer to __deepcopy__ https://github.com/search?q=repo%3Aprotocolbuffers%2Fprotobuf%20%22__deepcopy__%22&type=code , which calls MergeFrom; CopyFrom also uses MergeFrom: https://github.com/protocolbuffers/protobuf/blob/5076e6206e327942630e82bcec145826cf380559/python/google/protobuf/message.py#L91, so yes we can copy ourselves and save some cycles that performs all the branching.